File: app/src/main/java/com/hencoder/hencoderpracticedraw1/practice/Practice5DrawOvalView.java
package com.hencoder.hencoderpracticedraw1.practice;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

public class Practice5DrawOvalView extends View {

    private Paint mPaint;
    private int mLeftMargin, mTopMargin;
    private int mWidth = 400, mHeight = 200;
    private RectF mRectF;

    public Practice5DrawOvalView(Context context) {
        this(context, null);
    }

    public Practice5DrawOvalView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public Practice5DrawOvalView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mRectF = new RectF();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        mLeftMargin = (w - mWidth) / 2;
        mTopMargin = (h - mHeight) / 2;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        //        练习内容：使用 canvas.drawOval() 方法画椭圆

        mRectF.set(mLeftMargin, mTopMargin, mLeftMargin + mWidth, mTopMargin + mHeight);

        canvas.drawOval(mRectF, mPaint);

        //        canvas.drawArc(mLeftMargin, mTopMargin, mLeftMargin + mWidth, mTopMargin + mHeight, 0, 360, true, mPaint);

    }

}

File: app/src/main/java/com/hencoder/hencoderpracticedraw1/practice/Practice5DrawOvalView.kt
package com.hencoder.hencoderpracticedraw1.practice

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View

/**
 * Created by xh on 2017/10/16.
 */
class Practice5DrawOvalView : View {

    private val mPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val mRect = RectF()

    private var mLeftMargin = 0f
    private var mTopMargin = 0f

    private var mWidth = 400
    private var mHeight = 200

    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)

        mLeftMargin = (w - mWidth) / 2f
        mTopMargin = (h - mHeight) / 2f

        mRect.set(mLeftMargin, mTopMargin, mLeftMargin + mWidth, mTopMargin + mHeight)

    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        canvas!!.drawOval(mRect, mPaint)
    }

}